SlevomatCodingStandard/Sniffs/PHP/ForbiddenClassesSniff.php

```diff
@@ -0,0 +1,278 @@
+					$references,
+					$this->forbiddenTraits,
+					$tokens[$endTokenPointer]['code'] !== T_OPEN_CURLY_BRACKET
+				);
+			}
+		} elseif (in_array($token['code'], [T_NEW, T_EXTENDS], true)) {
+			$endTokenPointer = TokenHelper::findNextExcluding($phpcsFile, $nameTokens, $tokenPointer + 1);
+			$references = $this->getAllReferences($phpcsFile, $tokenPointer, $endTokenPointer);
+
+			$this->checkReferences(
+				$phpcsFile,
+				$tokenPointer,
+				$references,
+				$token['code'] === T_NEW ? $this->forbiddenClasses : $this->forbiddenExtends
+			);
+		} elseif ($token['code'] === T_DOUBLE_COLON && !$this->isTraitsConflictResolutionToken($token)) {
+			$startTokenPointer = TokenHelper::findPreviousExcluding($phpcsFile, $nameTokens, $tokenPointer - 1);
+			$references = $this->getAllReferences($phpcsFile, $startTokenPointer, $tokenPointer);
+
+			$this->checkReferences($phpcsFile, $tokenPointer, $references, $this->forbiddenClasses);
+		}
+	}
+
+	/**
+	 * @param File $phpcsFile
+	 * @param int $tokenPointer
+	 * @param array<stdClass> $references
+	 * @param array<string, (string|null)> $forbiddenNames
+	 * @param bool $isFixable
+	 */
+	private function checkReferences(
+		File $phpcsFile,
+		int $tokenPointer,
+		array $references,
+		array $forbiddenNames,
+		bool $isFixable = true
+	): void
+	{
+		$token = $phpcsFile->getTokens()[$tokenPointer];
+		$details = [
+			T_NEW => ['class', self::CODE_FORBIDDEN_CLASS],
+			T_DOUBLE_COLON => ['class', self::CODE_FORBIDDEN_CLASS],
+			T_EXTENDS => ['as a parent class', self::CODE_FORBIDDEN_PARENT_CLASS],
+			T_IMPLEMENTS => ['interface', self::CODE_FORBIDDEN_INTERFACE],
+			T_USE => ['trait', self::CODE_FORBIDDEN_TRAIT],
+		];
+
+		foreach ($references as $reference) {
+			if (!array_key_exists($reference->fullyQualifiedName, $forbiddenNames)) {
+				continue;
+			}
+
+			$alternative = $forbiddenNames[$reference->fullyQualifiedName];
+			[$nameType, $code] = $details[$token['code']];
+
+			if ($alternative === null) {
+				$phpcsFile->addError(
+					sprintf('Usage of %s %s is forbidden.', $reference->fullyQualifiedName, $nameType),
+					$reference->startPointer,
+					$code
+				);
+			} elseif (!$isFixable) {
+				$phpcsFile->addError(
+					sprintf(
+						'Usage of %s %s if forbidden, use %s instead.',
+						$reference->fullyQualifiedName,
+						$nameType,
+						$alternative
+					),
+					$reference->startPointer,
+					$code
+				);
+			} else {
+				$fix = $phpcsFile->addFixableError(
+					sprintf(
+						'Usage of %s %s if forbidden, use %s instead.',
+						$reference->fullyQualifiedName,
+						$nameType,
+						$alternative
+					),
+					$reference->startPointer,
+					$code
+				);
+				if (!$fix) {
+					continue;
+				}
+
+				$phpcsFile->fixer->beginChangeset();
+				$phpcsFile->fixer->replaceToken($reference->startPointer, $alternative);
+				for ($i = $reference->startPointer + 1; $i <= $reference->endPointer; ++$i) {
+					$phpcsFile->fixer->replaceToken((int) $i, '');
```

Comment:
`(int)` looks strange and should not be needed.

Reply:
I also thought, but then I got error from PHP Stan, that replace token expects `int`, but `int|float` is receiving. Will see if it disappear after rebase.

Reply:
I tried to remove `(int)` but this errors is shown: https://travis-ci.org/github/slevomat/coding-standard/jobs/672883870

Reply:
That's probably because of `min()` call and should be fixed there (via `@var` annotation)

Reply:
No it is not. I tried to put there `/** @var int $nextSeparator */` and errors remains.

Also if I change the line into this, which does not make sense, then PHP Stan pass. Only with the `$i` variable doesn't pass.

```php
// this works, when removed + 1 works also
$phpcsFile->fixer->replaceToken($reference->startPointer + 1, ''); 
```
I also tried to add `/** @var $i int */` into the for loop. This works, but PHP CS fails. with `No assignment to $i variable before or after the documentation comment.`

Reply:
Ok, another try:

Try to change `array<stdClass>` to array shape, see https://github.com/phpstan/phpdoc-parser/pull/30

So modify `stdClass` to array.

Reply:
Changing stdClass to an array helped. Please re-review